File: 0063-unique-paths-ii/0063-unique-paths-ii.py
class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        if obstacleGrid[0][0] == 1:
            return 0

        m, n = len(obstacleGrid), len(obstacleGrid[0])

        dp = [[0] * n for _ in range(m)]

        for col in range(n):
            if obstacleGrid[0][col] == 1:
                break
            dp[0][col] = 1

        for row in range(m):
            if obstacleGrid[row][0] == 1:
                break
            dp[row][0] = 1

        for row in range(1, m):
            for col in range(1, n):
                if obstacleGrid[row][col] != 1:
                    dp[row][col] += dp[row - 1][col] + dp[row][col - 1]

        return dp[-1][-1]

File: 0063-unique-paths-ii/0063-unique-paths-ii.cpp
class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
        if (obstacleGrid[0][0] == 1) {
            return 0;
        }

        if (obstacleGrid.size() == 1 && obstacleGrid[0].size() == 1) {
            return 1;
        }

        vector<vector<int>> dp(obstacleGrid.size(), vector<int>(obstacleGrid[0].size(), 0));


        for (size_t i = 1; i < obstacleGrid.size(); i++) {
            if (obstacleGrid[i][0] == 1) {
                break;
            }

            dp[i][0] = 1;
        }

        for (size_t i = 1; i < obstacleGrid[0].size(); i++) {
            if (obstacleGrid[0][i] == 1) {
                break;
            }
            
            dp[0][i] = 1;
        }

        for (size_t i = 1; i < obstacleGrid.size(); i++) {
            for (size_t j = 1; j < obstacleGrid[0].size(); j++) {
                if (obstacleGrid[i][j] == 1) {
                    continue;
                }

                dp[i][j] = dp[i - 1][j] + dp[i][j - 1];
            }
        }

        return dp[obstacleGrid.size() - 1] [obstacleGrid[0].size() - 1];
    }
};